Add helm chart for GCP CCM

eromanova opened the pull request:
Closes: https://github.com/k0rdent/kcm/issues/1174

eromanova committed:
Add helm chart for GCP CCM

Closes: https://github.com/k0rdent/kcm/issues/1174

zerospiel commented on mirantis/gcp-cloud-controller-manager/templates/clusterroles.yaml:
```diff
@@ -0,0 +1,161 @@
+  verbs:
+  - get
+  - update
+  - patch # until #393 lands
```
?

i believe that hasn't been fixed for the other charts.

eromanova replied:
Copy-pasting leftover. I've removed the comment.

zerospiel commented on mirantis/gcp-cloud-controller-manager/values.yaml:
```diff
@@ -0,0 +1,67 @@
+
+cloudConfig:
+  enabled: false
+  data: Cg== # base64-encoded cloud config
```
why do we need it?

eromanova replied:
We need to support passing custom GCP cloud config (don't mess with the GCP credentials, it's stored in another secret). The cloud config is also stored in the secret, and I added a `cloudConfig.data` field to support specifying the config in base64 encoded format.
After testing the CCM on our cloud, I verified that
```
[Global]
multizone=true
```
should be configured in the cloud config. So, we'll need to configure 
```
data: W0dsb2JhbF0KbXVsdGl6b25lPXRydWUK
```
for CCM to work properly. But it depends on the cloud configuration, so I thought it'll be good to support customizing the config.

`Cg==` indicates that this field should be base64 encoded.
Is it confusing?

zerospiel replied:
Sure, it makes sense to make it configurable. Not sure whether the comment will ever be read by any user, that is why I asked, and it is correct that having the `\n` b64 is a bit confusing. Can we keep it just empty with the comment? Otherwise it's okay to me

a13x5 commented on mirantis/gcp-cloud-controller-manager/templates/clusterrolebindings.yaml:
```diff
@@ -0,0 +1,26 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: system:controller:cloud-node-controller
```
Let's add common helm labels to all resources, since it's convention.
Also name should probably be unhardcoded and put to `_helpers` as a variable

eromanova replied:
No problem. I'll fix that. I just assumed the labels and annotations that Helm applies to all created resources would be sufficient.

a13x5 commented on mirantis/gcp-cloud-controller-manager/templates/cloudconfig.yaml:
```diff
@@ -0,0 +1,10 @@
+kind: Secret
+metadata:
+  name: {{ include "cloud-config-secret.name" . }}
+  namespace: kube-system
```
Namespace hardcode should be removed.
Applicable to all NS hardcode in the chart

eromanova replied:
The CCM driver does not work when it was installed in the namespace other than the `kube-system`. That's why I hardcoded it everywhere in the chart. It's done similarly in the Azure CCM helm chart. But I will double-check to be 100% sure, as I might have made a mistake during testing.

eromanova replied:
I've removed the hardcoded namespace for all the objects except API server reader rolebinding since the referenced role is automatically created in the `kube-system` namespace during GCP cluster creation. Added a comment with the explanation.

a13x5 commented on mirantis/gcp-cloud-controller-manager/templates/clusterrolebindings.yaml:
```diff
@@ -0,0 +1,26 @@
+- kind: ServiceAccount
+  apiGroup: ""
+  name: {{ .Values.serviceAccountName }}
+  namespace: kube-system
```
`{{ .Release.Namespace }}` ?
Same for other role bindings

zerospiel commented on mirantis/gcp-cloud-controller-manager/values.yaml:
```diff
@@ -0,0 +1,67 @@
+  repository: gcr.io/k8s-staging-cloud-provider-gcp/cloud-controller-manager
+  tag: v32.2.3
+
+# nameOverride overrides `cloud-controller-manager.fullname`
+nameOverride: "gcp-cloud-controller-manager"
```
according to the `_helpers.tpl` the comment is incorrect, only this value is used and there are no `fullname` or rather other default values, which is not obvious since the name of the variable contains `*override` hence it's implied to have some default value, isn't it?

a13x5 reviewed (changes requested):
NS hardcode
No helm labels
Names of resources are mostly hardcoded
Is there any reason for that?

zerospiel commented on mirantis/gcp-cloud-controller-manager/templates/clusterroles.yaml:
```diff
@@ -0,0 +1,161 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:controller:pvl-controller
```
out of curiosity: why is this role required, having no rolebinding for it? or do I miss something?

eromanova replied:
Yes, it seems to be a leftover from copy-pasting. Thanks for catching that.

zerospiel reviewed (commented):
minor comments

eromanova committed:
Address review comments

1. Unhardcoded the namespace and rbac objects' names (except api server
   reader rolebinding since the role is located in the kube-system
   namespace)
2. Added common helm labels on all resources
3. Updated the default cloud config value
4. Fixed several copy-pasting issues

zerospiel commented on mirantis/gcp-cloud-controller-manager/templates/clusterrolebindings.yaml:
```diff
@@ -0,0 +1,30 @@
+metadata:
+  name: {{ .Values.cloudNodeControllerRoleName }}
+  labels:
+    {{- include "gcp-ccm.labels" . | nindent 4 }}
```
super nit: when there is a new line like in the `labels:\n\t\t<>` there is no need in the `nindent` function call and vise-versa. applicable elsewhere. up to the PR author though

zerospiel commented on mirantis/gcp-cloud-controller-manager/templates/cloudconfig.yaml:
```diff
@@ -0,0 +1,11 @@
+  cloud-config: {{ .Values.cloudConfig.data }}
+kind: Secret
+metadata:
+  name: {{ include "cloud-config-secret.name" . }}
```
Generic question: why is the namespace field omitted everywhere for the ns-scoped resources?

zerospiel replied:
I see the comment [here](https://github.com/Mirantis/helm-charts/pull/3#discussion_r1989757472), but still it is unclear for me, AFAIK helm creates ns-scoped resources in the `default` ns and not in the `Release.Namespace`, so I am wondering whether it is the expected behavior

eromanova replied:
Helm will automatically set the namespace on all the objects to the `Release.Namespace` value, so I suggested omitting it as redundant.

zerospiel replied:
Uh, I haven't been aware of that. Alright then, thanks for the clarification!

zerospiel replied:
I cannot resolve the comments though, so do not hesitate to close them on your own

eromanova committed:
Fix CCM labels inclusion

a13x5 reviewed (approved):
LGTM
